File: cli/internal/wikis/parse.go
package wikis

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"path"
	"path/filepath"
	"sort"
	"strings"

	"harness-data/cli/internal/harness"
)

var allowedFrontmatter = map[string]bool{
	"name":             true,
	"label":            true,
	"aliases":          true,
	"negative_aliases": true,
	"covers":           true,
	"intents":          true,
	"status":           true,
	"object_type":      true,
	"canonical_status": true,
	"canonical_group":  true,
	"canonical_target": true,
	"canonical_reason": true,
}

type LoadCorpusOptions struct {
	FailFastParse bool
	ParseCodes    map[string]bool
}

func LoadCorpus(root string) (Corpus, []CheckError, error) {
	return LoadCorpusWithOptions(root, LoadCorpusOptions{})
}

func LoadCorpusWithOptions(root string, opts LoadCorpusOptions) (Corpus, []CheckError, error) {
	resolver, err := harness.NewPathResolver(root)
	if err != nil {
		return Corpus{}, nil, err
	}
	paths, err := collectCorpusMarkdown(resolver)
	if err != nil {
		return Corpus{}, nil, err
	}
	specSet := map[string]bool{}
	for _, p := range paths {
		if isSpecDocPath(p) {
			specSet[p] = true
		}
	}

	var docs []Document
	var errs []CheckError
	for _, logical := range paths {
		doc, parseErrs, err := ParseDocument(resolver, logical, specSet)
		if err != nil {
			return Corpus{}, nil, err
		}
		errs = append(errs, parseErrs...)
		docs = append(docs, doc)
		if opts.FailFastParse && hasSelectedParseErr(parseErrs, opts.ParseCodes) {
			break
		}
	}
	sort.Slice(docs, func(i, j int) bool { return docs[i].Path < docs[j].Path })
	byPath := map[string]*Document{}
	for i := range docs {
		byPath[docs[i].Path] = &docs[i]
	}
	return Corpus{Root: root, Docs: docs, ByPath: byPath, SpecPaths: specSet}, errs, nil
}

func hasSelectedParseErr(errs []CheckError, codes map[string]bool) bool {
	if len(errs) == 0 {
		return false
	}
	if len(codes) == 0 {
		return true
	}
	for _, err := range errs {
		if codes[err.Code] {
			return true
		}
	}
	return false
}

func ParseDocument(resolver harness.PathResolver, logical string, specSet map[string]bool) (Document, []CheckError, error) {
	physicalRel := resolver.ResolveRel(logical)
	data, err := os.ReadFile(filepath.Join(resolver.Root, filepath.FromSlash(physicalRel)))
	if err != nil {
		return Document{}, nil, err
	}
	fm, fmErrs := parseFrontmatter(logical, data)
	title, h1Count := parseH1(data)
	doc := Document{
		ID:             strings.TrimSuffix(logical, ".md"),
		Path:           logical,
		PhysicalRel:    physicalRel,
		Kind:           inferKind(logical),
		Domain:         inferDomain(logical),
		Title:          title,
		IsIndex:        path.Base(logical) == "index.md",
		HasFrontmatter: fm.Present,
	}
	if doc.Kind == KindSpec && !doc.IsIndex {
		doc.SpecType = inferSpecType(logical)
	}
	if doc.Kind == KindPlaybook && !doc.IsIndex {
		specPath := SamePath(logical, "spec")
		templatePath := SamePath(logical, "templates")
		doc.Playbook.SpecPath = specPath
		doc.Playbook.TemplatePath = templatePath
		if specSet[specPath] && inferSpecType(specPath) == SpecTypeMetric {
			doc.Playbook.IsSingle = true
		}
	}
	if doc.Kind == KindTemplate && !doc.IsIndex {
		doc.Template.PlaybookPath = SamePath(logical, "playbooks")
		doc.Template.IsReport = true
	}
	if fm.Present {
		doc.Name, _ = fm.Fields["name"].(string)
		doc.Label, _ = fm.Fields["label"].(string)
		doc.Aliases, _ = fm.Fields["aliases"].([]string)
		doc.NegativeAliases, _ = fm.Fields["negative_aliases"].([]string)
		doc.Covers, _ = fm.Fields["covers"].([]string)
		doc.Playbook.Intents, _ = fm.Fields["intents"].(map[string]PlaybookIntent)
	}
	errs := fmErrs
	if h1Count == 0 {
		errs = append(errs, CheckError{Path: logical, Code: "missing_h1", Message: "missing H1 title"})
	} else if h1Count > 1 {
		errs = append(errs, CheckError{Path: logical, Code: "multiple_h1", Message: "multiple H1 titles"})
	}
	return doc, errs, nil
}

func collectCorpusMarkdown(resolver harness.PathResolver) ([]string, error) {
	seen := map[string]bool{}
	var paths []string
	add := func(logical string) {
		if seen[logical] {
			return
		}
		seen[logical] = true
		paths = append(paths, logical)
	}
	if fileExists(resolver.Resolve("index.md")) {
		add("index.md")
	}
	for _, prefix := range []string{"spec", "playbooks", "templates", "metrics", "reports", "dims", "rules"} {
		found, err := collectLogicalMarkdown(resolver, prefix)
		if err != nil {
			return nil, err
		}
		for _, logical := range found {
			add(logical)
		}
	}
	sort.Strings(paths)
	return paths, nil
}

func collectLogicalMarkdown(resolver harness.PathResolver, prefix string) ([]string, error) {
	base := resolver.KnowledgePath(prefix)
	if _, err := os.Stat(base); os.IsNotExist(err) {
		return nil, nil
	}
	var paths []string
	err := filepath.WalkDir(base, func(file string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() || !strings.HasSuffix(d.Name(), ".md") {
			return nil
		}
		rel, err := filepath.Rel(resolver.Root, file)
		if err != nil {
			return err
		}
		paths = append(paths, resolver.LogicalRel(filepath.ToSlash(rel)))
		return nil
	})
	sort.Strings(paths)
	return paths, err
}

func inferKind(logical string) Kind {
	isIndex := path.Base(logical) == "index.md"
	switch {
	case logical == "index.md":
		return KindSpecIndex
	case isStructuredSpecPath(logical) && isIndex:
		return KindSpecIndex
	case isStructuredSpecPath(logical):
		return KindSpec
	case isStructuredPlaybookPath(logical):
		return KindPlaybook
	case isStructuredTemplatePath(logical):
		return KindTemplate
	case strings.HasPrefix(logical, "spec/") && isIndex:
		return KindSpecIndex
	case strings.HasPrefix(logical, "spec/"):
		return KindSpec
	case strings.HasPrefix(logical, "playbooks/") && isIndex:
		return KindPlaybookIndex
	case strings.HasPrefix(logical, "playbooks/"):
		return KindPlaybook
	case strings.HasPrefix(logical, "templates/") && isIndex:
		return KindTemplateIndex
	default:
		return KindTemplate
	}
}

func inferDomain(logical string) string {
	if logical == "index.md" {
		return ""
	}
	if isStructuredPath(logical) {
		parts := strings.Split(logical, "/")
		if len(parts) == 2 && parts[1] == "index.md" {
			return ""
		}
		if len(parts) >= 2 {
			return parts[1]
		}
		return ""
	}
	parts := strings.SplitN(logical, "/", 2)
	if len(parts) != 2 {
		return ""
	}
	rest := parts[1]
	if path.Base(rest) == "index.md" {
		domain := path.Dir(rest)
		if domain == "." {
			return ""
		}
		return domain
	}
	domain := path.Dir(rest)
	if domain == "." {
		return ""
	}
	return domain
}

func SamePath(logical, targetRoot string) string {
	if isStructuredPath(logical) {
		dir := path.Dir(logical)
		switch targetRoot {
		case "spec":
			return path.Join(dir, "spec.md")
		case "playbooks":
			return path.Join(dir, "playbook.md")
		case "templates":
			return path.Join(dir, "template.md")
		default:
			return logical
		}
	}
	parts := strings.SplitN(logical, "/", 2)
	if len(parts) != 2 {
		return logical
	}
	return targetRoot + "/" + parts[1]
}

func isSpecDocPath(logical string) bool {
	return inferKind(logical) == KindSpec && path.Base(logical) != "index.md"
}

func inferSpecType(logical string) SpecType {
	if strings.HasPrefix(logical, "metrics/") {
		return SpecTypeMetric
	}
	if strings.HasPrefix(logical, "reports/") || strings.HasPrefix(logical, "dims/") || strings.HasPrefix(logical, "rules/") {
		return SpecTypeConcept
	}
	base := path.Base(logical)
	if strings.HasPrefix(base, "c-") || strings.HasPrefix(base, "r-") {
		return SpecTypeConcept
	}
	return SpecTypeMetric
}

func isStructuredPath(logical string) bool {
	return strings.HasPrefix(logical, "metrics/") ||
		strings.HasPrefix(logical, "reports/") ||
		strings.HasPrefix(logical, "dims/") ||
		strings.HasPrefix(logical, "rules/")
}

func isStructuredSpecPath(logical string) bool {
	base := path.Base(logical)
	if base == "spec.md" {
		return isStructuredPath(logical)
	}
	return base == "index.md" && isStructuredPath(logical)
}

func isStructuredPlaybookPath(logical string) bool {
	return isStructuredPath(logical) && path.Base(logical) == "playbook.md"
}

func isStructuredTemplatePath(logical string) bool {
	return isStructuredPath(logical) && path.Base(logical) == "template.md"
}

func parseH1(data []byte) (string, int) {
	scanner := bufio.NewScanner(bytes.NewReader(stripFrontmatter(data)))
	count := 0
	title := ""
	inFence := false
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if strings.HasPrefix(line, "```") || strings.HasPrefix(line, "~~~") {
			inFence = !inFence
			continue
		}
		if inFence {
			continue
		}
		if strings.HasPrefix(line, "# ") {
			count++
			if title == "" {
				title = strings.TrimSpace(strings.TrimPrefix(line, "# "))
			}
		}
	}
	return title, count
}

func stripFrontmatter(data []byte) []byte {
	lines := bytes.Split(data, []byte("\n"))
	if len(lines) == 0 || strings.TrimSpace(string(lines[0])) != "---" {
		return data
	}
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(string(lines[i])) == "---" {
			return bytes.Join(lines[i+1:], []byte("\n"))
		}
	}
	return data
}

func parseFrontmatter(logical string, data []byte) (Frontmatter, []CheckError) {
	lines := strings.Split(string(data), "\n")
	if len(lines) == 0 || strings.TrimSpace(lines[0]) != "---" {
		return Frontmatter{Fields: map[string]any{}}, nil
	}
	end := -1
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == "---" {
			end = i
			break
		}
	}
	if end == -1 {
		return Frontmatter{Present: true, Fields: map[string]any{}}, []CheckError{{Path: logical, Code: "invalid_frontmatter_type", Message: "unterminated frontmatter"}}
	}
	fm := Frontmatter{Present: true, Fields: map[string]any{}}
	var errs []CheckError
	for i := 1; i < end; i++ {
		raw := strings.TrimRight(lines[i], " \t")
		trimmed := strings.TrimSpace(raw)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		if strings.HasPrefix(trimmed, "- ") {
			continue
		}
		indent := len(raw) - len(strings.TrimLeft(raw, " "))
		if indent != 0 {
			continue
		}
		key, value, ok := strings.Cut(trimmed, ":")
		if !ok {
			errs = append(errs, CheckError{Path: logical, Code: "invalid_frontmatter_type", Message: fmt.Sprintf("invalid frontmatter line: %s", trimmed)})
			continue
		}
		key = strings.TrimSpace(key)
		value = strings.TrimSpace(value)
		if !allowedFrontmatter[key] {
			errs = append(errs, CheckError{Path: logical, Code: "unknown_frontmatter_field", Message: "unknown frontmatter field", Target: key})
			continue
		}
		switch key {
		case "name", "label", "status", "object_type", "canonical_status", "canonical_group", "canonical_target", "canonical_reason":
			if value == "" || strings.HasPrefix(value, "[") {
				errs = append(errs, CheckError{Path: logical, Code: "invalid_frontmatter_type", Message: "frontmatter field must be a string", Target: key})
				continue
			}
			fm.Fields[key] = cleanScalar(value)
		case "aliases", "negative_aliases", "covers":
			values, ok := parseStringArray(lines, &i, end, value)
			if !ok {
				code := "invalid_frontmatter_type"
				if key == "covers" {
					code = "invalid_covers_type"
				}
				errs = append(errs, CheckError{Path: logical, Code: code, Message: "frontmatter field must be a string array", Target: key})
				continue
			}
			fm.Fields[key] = values
		case "intents":
			intents, ok := parsePlaybookIntents(lines, &i, end, value)
			if !ok {
				errs = append(errs, CheckError{Path: logical, Code: "invalid_frontmatter_type", Message: "frontmatter field must be an intents map", Target: key})
				continue
			}
			fm.Fields[key] = intents
		}
	}
	return fm, errs
}

func parsePlaybookIntents(lines []string, idx *int, end int, value string) (map[string]PlaybookIntent, bool) {
	if value != "" {
		return nil, false
	}
	out := map[string]PlaybookIntent{}
	for *idx+1 < end {
		next := lines[*idx+1]
		trimmed := strings.TrimSpace(next)
		if trimmed == "" {
			(*idx)++
			continue
		}
		indent := len(next) - len(strings.TrimLeft(next, " "))
		if indent == 0 {
			break
		}
		if indent != 2 {
			return nil, false
		}
		key, nestedValue, ok := strings.Cut(trimmed, ":")
		if !ok || strings.TrimSpace(key) == "" || strings.TrimSpace(nestedValue) != "" {
			return nil, false
		}
		intentName := cleanScalar(key)
		(*idx)++
		intent := PlaybookIntent{}
		sawField := false
		for *idx+1 < end {
			fieldLine := lines[*idx+1]
			fieldTrimmed := strings.TrimSpace(fieldLine)
			if fieldTrimmed == "" {
				(*idx)++
				continue
			}
			fieldIndent := len(fieldLine) - len(strings.TrimLeft(fieldLine, " "))
			if fieldIndent <= 2 {
				break
			}
			if fieldIndent != 4 {
				return nil, false
			}
			fieldKey, fieldValue, ok := strings.Cut(fieldTrimmed, ":")
			if !ok {
				return nil, false
			}
			fieldKey = strings.TrimSpace(fieldKey)
			fieldValue = strings.TrimSpace(fieldValue)
			if fieldKey != "aliases" {
				return nil, false
			}
			values, ok := parseIndentedStringArray(lines, idx, end, fieldValue, fieldIndent)
			if !ok {
				return nil, false
			}
			intent.Aliases = values
			sawField = true
		}
		if !sawField || intent.Aliases == nil {
			return nil, false
		}
		out[intentName] = intent
	}
	return out, true
}

func parseIndentedStringArray(lines []string, idx *int, end int, value string, parentIndent int) ([]string, bool) {
	if strings.HasPrefix(value, "[") {
		if !strings.HasSuffix(value, "]") {
			return nil, false
		}
		inner := strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(value, "["), "]"))
		if inner == "" {
			(*idx)++
			return []string{}, true
		}
		parts := strings.Split(inner, ",")
		out := make([]string, 0, len(parts))
		for _, part := range parts {
			out = append(out, cleanScalar(part))
		}
		(*idx)++
		return out, true
	}
	if value != "" {
		return nil, false
	}
	(*idx)++
	var out []string
	for *idx+1 < end {
		next := lines[*idx+1]
		trimmed := strings.TrimSpace(next)
		if trimmed == "" {
			(*idx)++
			continue
		}
		indent := len(next) - len(strings.TrimLeft(next, " "))
		if indent <= parentIndent {
			break
		}
		if !strings.HasPrefix(trimmed, "- ") {
			return nil, false
		}
		out = append(out, cleanScalar(strings.TrimPrefix(trimmed, "- ")))
		(*idx)++
	}
	return out, true
}

func parseStringArray(lines []string, idx *int, end int, value string) ([]string, bool) {
	if strings.HasPrefix(value, "[") {
		if !strings.HasSuffix(value, "]") {
			return nil, false
		}
		inner := strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(value, "["), "]"))
		if inner == "" {
			return []string{}, true
		}
		parts := strings.Split(inner, ",")
		out := make([]string, 0, len(parts))
		for _, part := range parts {
			out = append(out, cleanScalar(part))
		}
		return out, true
	}
	if value != "" {
		return nil, false
	}
	var out []string
	for *idx+1 < end {
		next := lines[*idx+1]
		trimmed := strings.TrimSpace(next)
		if trimmed == "" {
			(*idx)++
			continue
		}
		indent := len(next) - len(strings.TrimLeft(next, " "))
		if indent == 0 {
			break
		}
		if !strings.HasPrefix(trimmed, "- ") {
			return nil, false
		}
		out = append(out, cleanScalar(strings.TrimPrefix(trimmed, "- ")))
		(*idx)++
	}
	return out, true
}

func cleanScalar(s string) string {
	s = strings.TrimSpace(s)
	if hash := strings.Index(s, " #"); hash >= 0 {
		s = strings.TrimSpace(s[:hash])
	}
	return strings.Trim(s, `"'`)
}

File: packages/data-harness-cli/src/lib/wikis/parse.js
import { readdirSync, readFileSync, statSync } from "node:fs";
import path from "node:path";

import { newPathResolver } from "../harness.js";
import {
  isStructuredPath,
  KIND_PLAYBOOK,
  KIND_SPEC,
  KIND_TEMPLATE,
  samePath,
  SPEC_TYPE_CONCEPT,
  SPEC_TYPE_METRIC,
} from "./paths.js";

const ALLOWED_FRONTMATTER = new Set([
  "name",
  "label",
  "aliases",
  "negative_aliases",
  "covers",
  "intents",
  "canonical_status",
  "canonical_group",
  "canonical_target",
  "canonical_reason",
]);

export const KIND_SPEC_INDEX = "spec_index";
export const KIND_PLAYBOOK_INDEX = "playbook_index";
export const KIND_TEMPLATE_INDEX = "template_index";

export function loadCorpus(root) {
  return loadCorpusWithOptions(root, {});
}

export function loadCorpusWithOptions(root, opts = {}) {
  const resolver = newPathResolver(root);
  const paths = collectCorpusMarkdown(resolver);
  const specSet = {};
  for (const p of paths) {
    if (isSpecDocPath(p)) specSet[p] = true;
  }
  const docs = [];
  const errs = [];
  for (const logical of paths) {
    const parsed = parseDocument(resolver, logical, specSet);
    errs.push(...parsed.errs);
    docs.push(parsed.doc);
    if (opts.failFastParse && hasSelectedParseErr(parsed.errs, opts.parseCodes)) break;
  }
  docs.sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0));
  const byPath = Object.fromEntries(docs.map((doc) => [doc.path, doc]));
  return { corpus: { root, docs, byPath, specPaths: specSet }, errs };
}

function hasSelectedParseErr(errs, codes) {
  if (!errs.length) return false;
  if (!codes) return true;
  return errs.some((err) => codes.has(err.code));
}

export function parseDocument(resolver, logical, specSet) {
  const physicalRel = resolver.resolveRel(logical);
  const data = readFileSync(path.join(resolver.root, physicalRel));
  const { fm, errs: fmErrs } = parseFrontmatter(logical, data.toString("utf8"));
  const { title, h1Count } = parseH1(data);
  const doc = {
    id: logical.replace(/\.md$/, ""),
    path: logical,
    physicalRel,
    kind: inferKind(logical),
    domain: inferDomain(logical),
    title,
    isIndex: path.posix.basename(logical) === "index.md",
    hasFrontmatter: fm.present,
    playbook: {},
    template: {},
  };
  if (doc.kind === KIND_SPEC && !doc.isIndex) doc.specType = inferSpecType(logical);
  if (doc.kind === KIND_PLAYBOOK && !doc.isIndex) {
    const specPath = samePath(logical, "spec");
    const templatePath = samePath(logical, "templates");
    doc.playbook.specPath = specPath;
    doc.playbook.templatePath = templatePath;
    if (specSet[specPath] && inferSpecType(specPath) === SPEC_TYPE_METRIC) doc.playbook.isSingle = true;
  }
  if (doc.kind === KIND_TEMPLATE && !doc.isIndex) {
    doc.template.playbookPath = samePath(logical, "playbooks");
    doc.template.isReport = true;
  }
  if (fm.present) {
    if (typeof fm.fields.name === "string") doc.name = fm.fields.name;
    if (typeof fm.fields.label === "string") doc.label = fm.fields.label;
    if (Array.isArray(fm.fields.aliases)) doc.aliases = fm.fields.aliases;
    if (Array.isArray(fm.fields.negative_aliases)) doc.negativeAliases = fm.fields.negative_aliases;
    if (Array.isArray(fm.fields.covers)) doc.covers = fm.fields.covers;
    if (fm.fields.intents) doc.playbook.intents = fm.fields.intents;
  }
  const errs = [...fmErrs];
  if (h1Count === 0) errs.push({ path: logical, code: "missing_h1", message: "missing H1 title" });
  else if (h1Count > 1) errs.push({ path: logical, code: "multiple_h1", message: "multiple H1 titles" });
  return { doc, errs };
}

function collectCorpusMarkdown(resolver) {
  const seen = new Set();
  const paths = [];
  const add = (logical) => {
    if (seen.has(logical)) return;
    seen.add(logical);
    paths.push(logical);
  };
  try {
    if (statSync(resolver.resolve("index.md")).isFile()) add("index.md");
  } catch {
    // missing
  }
  for (const prefix of ["spec", "playbooks", "templates", "metrics", "reports", "dims", "rules"]) {
    for (const logical of collectLogicalMarkdown(resolver, prefix)) add(logical);
  }
  paths.sort();
  return paths;
}

function collectLogicalMarkdown(resolver, prefix) {
  const base = resolver.knowledgePath(prefix);
  try {
    if (!statSync(base).isDirectory()) return [];
  } catch {
    return [];
  }
  const paths = [];
  walkMarkdown(base, resolver.root, resolver, paths);
  paths.sort();
  return paths;
}

function walkMarkdown(dir, root, resolver, paths) {
  for (const name of readdirSync(dir)) {
    const file = path.join(dir, name);
    const info = statSync(file);
    if (info.isDirectory()) {
      walkMarkdown(file, root, resolver, paths);
      continue;
    }
    if (!name.endsWith(".md")) continue;
    const rel = path.relative(root, file);
    paths.push(resolver.logicalRel(rel.split(path.sep).join("/")));
  }
}

export function inferKind(logical) {
  const isIndex = path.posix.basename(logical) === "index.md";
  if (logical === "index.md") return KIND_SPEC_INDEX;
  if (isStructuredSpecPath(logical) && isIndex) return KIND_SPEC_INDEX;
  if (isStructuredSpecPath(logical)) return KIND_SPEC;
  if (isStructuredPlaybookPath(logical)) return KIND_PLAYBOOK;
  if (isStructuredTemplatePath(logical)) return KIND_TEMPLATE;
  if (logical.startsWith("spec/") && isIndex) return KIND_SPEC_INDEX;
  if (logical.startsWith("spec/")) return KIND_SPEC;
  if (logical.startsWith("playbooks/") && isIndex) return KIND_PLAYBOOK_INDEX;
  if (logical.startsWith("playbooks/")) return KIND_PLAYBOOK;
  if (logical.startsWith("templates/") && isIndex) return KIND_TEMPLATE_INDEX;
  return KIND_TEMPLATE;
}

function inferDomain(logical) {
  if (logical === "index.md") return "";
  if (isStructuredPath(logical)) {
    const parts = logical.split("/");
    if (parts.length === 2 && parts[1] === "index.md") return "";
    if (parts.length >= 2) return parts[1];
    return "";
  }
  const slash = logical.indexOf("/");
  if (slash < 0) return "";
  const rest = logical.slice(slash + 1);
  if (path.posix.basename(rest) === "index.md") {
    const domain = path.posix.dirname(rest);
    return domain === "." ? "" : domain;
  }
  const domain = path.posix.dirname(rest);
  return domain === "." ? "" : domain;
}

export function isSpecDocPath(logical) {
  return inferKind(logical) === KIND_SPEC && path.posix.basename(logical) !== "index.md";
}

export function inferSpecType(logical) {
  if (logical.startsWith("metrics/")) return SPEC_TYPE_METRIC;
  if (logical.startsWith("reports/") || logical.startsWith("dims/") || logical.startsWith("rules/")) return SPEC_TYPE_CONCEPT;
  const base = path.posix.basename(logical);
  if (base.startsWith("c-") || base.startsWith("r-")) return SPEC_TYPE_CONCEPT;
  return SPEC_TYPE_METRIC;
}

function isStructuredSpecPath(logical) {
  const base = path.posix.basename(logical);
  if (base === "spec.md") return isStructuredPath(logical);
  return base === "index.md" && isStructuredPath(logical);
}

function isStructuredPlaybookPath(logical) {
  return isStructuredPath(logical) && path.posix.basename(logical) === "playbook.md";
}

function isStructuredTemplatePath(logical) {
  return isStructuredPath(logical) && path.posix.basename(logical) === "template.md";
}

function parseH1(data) {
  const text = stripFrontmatter(data.toString("utf8"));
  let count = 0;
  let title = "";
  for (const line of text.split(/\n/)) {
    const trimmed = line.trim();
    if (trimmed.startsWith("# ")) {
      count += 1;
      if (!title) title = trimmed.slice(2).trim();
    }
  }
  return { title, h1Count: count };
}

function stripFrontmatter(text) {
  const lines = text.split("\n");
  if (!lines.length || lines[0].trim() !== "---") return text;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === "---") return lines.slice(i + 1).join("\n");
  }
  return text;
}

export function frontmatterAliasFields(text) {
  const { fm } = parseFrontmatter("", text);
  return {
    aliases: Array.isArray(fm.fields.aliases) ? fm.fields.aliases : [],
    negativeAliases: Array.isArray(fm.fields.negative_aliases) ? fm.fields.negative_aliases : [],
  };
}

export function stripFrontmatterText(text) {
  return stripFrontmatter(text);
}

function parseFrontmatter(logical, text) {
  const lines = text.split("\n");
  if (!lines.length || lines[0].trim() !== "---") return { fm: { present: false, fields: {} }, errs: [] };
  let end = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === "---") {
      end = i;
      break;
    }
  }
  if (end === -1) {
    return {
      fm: { present: true, fields: {} },
      errs: [{ path: logical, code: "invalid_frontmatter_type", message: "unterminated frontmatter" }],
    };
  }
  const fm = { present: true, fields: {} };
  const errs = [];
  for (let i = 1; i < end; i++) {
    const raw = lines[i].replace(/[ \t]+$/, "");
    const trimmed = raw.trim();
    if (!trimmed || trimmed.startsWith("#") || trimmed.startsWith("- ")) continue;
    const indent = raw.length - raw.trimStart().length;
    if (indent !== 0) continue;
    const cut = trimmed.indexOf(":");
    if (cut < 0) {
      errs.push({ path: logical, code: "invalid_frontmatter_type", message: `invalid frontmatter line: ${trimmed}` });
      continue;
    }
    const key = trimmed.slice(0, cut).trim();
    const value = trimmed.slice(cut + 1).trim();
    if (!ALLOWED_FRONTMATTER.has(key)) {
      errs.push({ path: logical, code: "unknown_frontmatter_field", message: "unknown frontmatter field", target: key });
      continue;
    }
    const cursor = { i };
    if (["name", "label", "canonical_status", "canonical_group", "canonical_target", "canonical_reason"].includes(key)) {
      if (!value || value.startsWith("[")) {
        errs.push({ path: logical, code: "invalid_frontmatter_type", message: "frontmatter field must be a string", target: key });
        continue;
      }
      fm.fields[key] = cleanScalar(value);
    } else if (key === "aliases" || key === "negative_aliases" || key === "covers") {
      const values = parseStringArray(lines, cursor, end, value);
      if (!values) {
        errs.push({
          path: logical,
          code: key === "covers" ? "invalid_covers_type" : "invalid_frontmatter_type",
          message: "frontmatter field must be a string array",
          target: key,
        });
        continue;
      }
      fm.fields[key] = values;
      i = cursor.i;
    } else if (key === "intents") {
      const intents = parsePlaybookIntents(lines, cursor, end, value);
      if (!intents) {
        errs.push({ path: logical, code: "invalid_frontmatter_type", message: "frontmatter field must be an intents map", target: key });
        continue;
      }
      fm.fields[key] = intents;
      i = cursor.i;
    }
  }
  return { fm, errs };
}

function parsePlaybookIntents(lines, cursor, end, value) {
  if (value !== "") return null;
  const out = {};
  while (cursor.i + 1 < end) {
    const next = lines[cursor.i + 1];
    const trimmed = next.trim();
    if (!trimmed) {
      cursor.i += 1;
      continue;
    }
    const indent = next.length - next.trimStart().length;
    if (indent === 0) break;
    if (indent !== 2) return null;
    const cut = trimmed.indexOf(":");
    if (cut < 0 || !trimmed.slice(0, cut).trim() || trimmed.slice(cut + 1).trim() !== "") return null;
    const intentName = cleanScalar(trimmed.slice(0, cut));
    cursor.i += 1;
    const intent = { aliases: null };
    let sawField = false;
    while (cursor.i + 1 < end) {
      const fieldLine = lines[cursor.i + 1];
      const fieldTrimmed = fieldLine.trim();
      if (!fieldTrimmed) {
        cursor.i += 1;
        continue;
      }
      const fieldIndent = fieldLine.length - fieldLine.trimStart().length;
      if (fieldIndent <= 2) break;
      if (fieldIndent !== 4) return null;
      const fieldCut = fieldTrimmed.indexOf(":");
      if (fieldCut < 0) return null;
      const fieldKey = fieldTrimmed.slice(0, fieldCut).trim();
      const fieldValue = fieldTrimmed.slice(fieldCut + 1).trim();
      if (fieldKey !== "aliases") return null;
      const values = parseIndentedStringArray(lines, cursor, end, fieldValue, fieldIndent);
      if (!values) return null;
      intent.aliases = values;
      sawField = true;
    }
    if (!sawField || intent.aliases == null) return null;
    out[intentName] = intent;
  }
  return out;
}

function parseIndentedStringArray(lines, cursor, end, value, parentIndent) {
  if (value.startsWith("[")) {
    if (!value.endsWith("]")) return null;
    const inner = value.slice(1, -1).trim();
    cursor.i += 1;
    if (!inner) return [];
    return inner.split(",").map((part) => cleanScalar(part));
  }
  if (value !== "") return null;
  cursor.i += 1;
  const out = [];
  while (cursor.i + 1 < end) {
    const next = lines[cursor.i + 1];
    const trimmed = next.trim();
    if (!trimmed) {
      cursor.i += 1;
      continue;
    }
    const indent = next.length - next.trimStart().length;
    if (indent <= parentIndent) break;
    if (!trimmed.startsWith("- ")) return null;
    out.push(cleanScalar(trimmed.slice(2)));
    cursor.i += 1;
  }
  return out;
}

function parseStringArray(lines, cursor, end, value) {
  if (value.startsWith("[")) {
    if (!value.endsWith("]")) return null;
    const inner = value.slice(1, -1).trim();
    if (!inner) return [];
    return inner.split(",").map((part) => cleanScalar(part));
  }
  if (value !== "") return null;
  const out = [];
  while (cursor.i + 1 < end) {
    const next = lines[cursor.i + 1];
    const trimmed = next.trim();
    if (!trimmed) {
      cursor.i += 1;
      continue;
    }
    const indent = next.length - next.trimStart().length;
    if (indent === 0) break;
    if (!trimmed.startsWith("- ")) return null;
    out.push(cleanScalar(trimmed.slice(2)));
    cursor.i += 1;
  }
  return out;
}

function cleanScalar(s) {
  s = s.trim();
  const hash = s.indexOf(" #");
  if (hash >= 0) s = s.slice(0, hash).trim();
  return s.replace(/^["']|["']$/g, "");
}
